src/core/atomic.d

```diff
@@ -32,12 +32,65 @@ else
 
 private
 {
-    template HeadUnshared(T)
+    /* Construct a type with a shared tail, and if possible with an unshared
+    head. */
+    template TailShared(T)
     {
-        static if( is( T U : shared(U*) ) )
-            alias shared(U)* HeadUnshared;
+        alias S = shared T;
+
+        static if (is(S U == shared U)) {}
+        else static assert(false, "Should never be triggered. The `static " ~
+            "if` declares `U` as the unshared version of the shared type " ~
+            "`S`. `S` is explicitly declared as shared, so getting `U` " ~
+            "should always work.");
```

Comment:
> S is defined as shared T so this static if essentially extracts the T back under the name U. What is the purpose?

U is different from T when T is shared. T may be shared or not. S is always shared. U is always unshared.